### PR TITLE
feature/jl/5.5_add_select_users_permission

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: sonar-db/src/main/java/org/sonar/db/permission/PermissionMapper.java [hunk 1]
```diff
@@ -30,6 +31,8 @@
 
   int countUsers(Map<String, Object> parameters);
 
+  List<UserWithPermissionDto> selectUsersByPermissionAndComponent(@Param("permission") String permission, @Param("componentId") long componentId, @Param("userIds") List<Long> userIds);
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/permission/PermissionMapper.java]: MINOR Split this 184 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00103)

File: sonar-db/src/main/java/org/sonar/db/permission/PermissionDao.java [hunk 2]
```diff
@@ -143,4 +150,21 @@ public Void apply(@Nonnull List<Long> partitionedComponentIds) {
...
+    private final long componentId;
+
+    private SelectUsersByPermissionAndComponent(PermissionMapper mapper, String permission, long componentId) {
+      this.mapper = mapper;
+      this.permission = permission;
+      this.componentId = componentId;
+    }
+
+    @Override
+    public List<UserWithPermissionDto> apply(@Nonnull List<Long> partitionOfIds) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/permission/PermissionDao.java]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: sonar-db/src/main/java/org/sonar/db/user/AuthorizationDao.java [hunk 3]
```diff
@@ -129,4 +137,21 @@ public boolean isAuthorizedComponentKey(String componentKey, @Nullable Integer u
...
+    private final long componentId;
+
+    private SelectUsersByPermissionAndProject(AuthorizationMapper mapper, String role, long componentId) {
+      this.mapper = mapper;
+      this.role = role;
+      this.componentId = componentId;
+    }
+
+    @Override
+    public List<Long> apply(@Nonnull List<Long> partitionOfIds) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/user/AuthorizationDao.java]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: sonar-db/src/main/java/org/sonar/db/user/AuthorizationDao.java [hunk 4]
```diff
@@ -129,4 +137,21 @@ public boolean isAuthorizedComponentKey(String componentKey, @Nullable Integer u
       MyBatis.closeQuietly(session);
     }
   }
+
+  private static class SelectUsersByPermissionAndProject implements Function<List<Long>, List<Long>> {
+    private final AuthorizationMapper mapper;
+    private final String role;
+    private final long componentId;
+
+    private SelectUsersByPermissionAndProject(AuthorizationMapper mapper, String role, long componentId) {
```
Comment (simonbrandhof) [inline sonar-db/src/main/java/org/sonar/db/user/AuthorizationDao.java]: I suggest to rename `componentId` to `projectId` as it only accepts projects.

Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/user/AuthorizationDao.java]: Indeed (in fact it also accepts views, but projectId seems ok for me)

**Comment by simonbrandhof** [comment]
LGTM

**Pull request merged**